libs/wallet/src/connectors/json-rpc-connector.ts

```diff
@@ -91,116 +28,104 @@ export class JsonRpcConnector implements VegaConnector {
   token: string | null = null;
   reqId = 0;
+  client?: WalletClient;
 
   constructor() {
     const cfg = getConfig();
-    if (cfg) {
+
+    if (cfg && cfg.url) {
       this.token = cfg.token;
       this.url = cfg.url;
+      this.client = new WalletClient({
+        hostname: cfg.url,
+        token: cfg.token ?? undefined,
+      });
     }
   }
 
   async getChainId() {
-    const result = await this.request(Methods.GetChainId);
-    if ('error' in result) {
-      throw this.wrapError(result.error);
+    if (!this.client) {
+      throw new Error(
+        'No client found. The connector needs to be initialized with a url.'
+      );
     }
-    const parseResult = GetChainIdSchema.safeParse(result);
-    if (parseResult.success) {
-      return parseResult.data.result;
-    } else {
+    try {
+      const { result } = await this.client.GetChainId();
+      return result;
+    } catch (err) {
       throw ClientErrors.INVALID_RESPONSE;
     }
   }
 
   async connectWallet() {
-    const result = await this.request(Methods.ConnectWallet, {
-      hostname: window.location.host,
-    });
-    if ('error' in result) {
-      throw this.wrapError(result.error);
+    if (!this.client) {
+      throw new Error(
+        'No client found. The connector needs to be initialized with a url.'
+      );
     }
-    const parseResult = ConnectWalletSchema.safeParse(result);
 
-    if (parseResult.success) {
-      // store token and other config for eager connect and subsequent requests
+    try {
+      const { result } = await this.client.ConnectWallet();
       setConfig({
-        token: parseResult.data.result.token,
+        token: result.token,
         connector: 'jsonRpc',
         url: this.url,
       });
-      return parseResult.data.result;
-    } else {
+      return result;
+    } catch (err) {
       throw ClientErrors.INVALID_RESPONSE;
     }
   }
 
   // connect actually calling list_keys here, not to be confused with connect_wallet
   // which retrieves the session token
   async connect() {
-    const cfg = getConfig();
-    if (!cfg?.token) {
-      throw ClientErrors.NO_TOKEN;
-    }
-    const result = await this.request(Methods.ListKeys, {
-      token: cfg.token,
-    });
-    if ('error' in result) {
-      throw this.wrapError(result.error);
+    if (!this.client) {
+      throw new Error(
+        'No client found. The connector needs to be initialized with a url.'
+      );
     }
-    const parseResult = ListKeysSchema.safeParse(result);
-    if (parseResult.success) {
-      return parseResult.data.result.keys;
-    } else {
+
+    try {
+      const { result } = await this.client.ListKeys({});
+      return result.keys;
+    } catch (err) {
       throw ClientErrors.INVALID_RESPONSE;
     }
   }
 
   async disconnect() {
-    const cfg = getConfig();
-
-    if (cfg?.token) {
-      await this.request(Methods.DisconnectWallet, {
-        token: cfg.token,
-      });
+    if (!this.client) {
+      throw new Error(
```

Comment:
chore (non blocking): This could be DRY'd up a bit. Perhaps with a method `this.checkClient()` to check this.client is set? or storing the error as a static property?

Reply:
1. i could implement a checkClient, but it wouldn't add any value in typescript. having a difficult time explaining it with words on a monday morning, but:

```
class JsonRpcConnector {
  checkClient() {
     if (!this.client) {
       throw new Error()
     }
   }
   
   someMethod() {
     this.checkClient()
     this.client.SomeMethod() // <= Error: Object possibly undefined
   }

}
```

this could be all scraped if the url was a required prop of the connector, but i'd need to rewrite a while bunch of things for that, and i was just not up for the challenge, but i think it would be a much nicer approach than having these checks for undefined all over the place. 

2. the error as a static property i think in general a not so good of a practice, because it totally swallows the stack trace, so when debugging, you wouldn't be able to identify which actual method threw the error. stack traces get injected upon initialisation, not a huge expert on the topic, but did the following little test to confirm:

```
// 1 - storing the pre-initialised error
const e = new Error('test stored')
function f1 () {
  throw e
}


// 2 - not storing it
function f2 {
  throw new Error('test not stored')
}
```

The result when calling them looks different:

<img width="284" alt="Screenshot 2022-12-26 at 11 32 48" src="https://user-images.githubusercontent.com/105208209/209544404-49e7192c-bd08-4a53-9698-c39737a09c32.png">

Reply:
Makes a lot of sense. Thanks for the detailed response